Source/ARTJsonEncoder.m

```diff
@@ -22,12 +22,20 @@ - (NSString *)formatAsString {
     return @"json";
 }
 
-- (id)decode:(NSData *)data {
-    return [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
+- (id)decode:(NSData *)data error:(NSError **)error {
+    return [NSJSONSerialization JSONObjectWithData:data options:0 error:error];
 }
 
-- (NSData *)encode:(id)obj {
-    return [NSJSONSerialization dataWithJSONObject:obj options:0 error:nil];
+- (NSData *)encode:(id)obj error:(NSError **)error {
+    @try {
+        return [NSJSONSerialization dataWithJSONObject:obj options:0 error:error];
+    }
+    @catch (NSException *exception) {
+        if (error) {
+            *error = [[NSError alloc] initWithDomain:ARTAblyErrorDomain code:ARTClientCodeErrorInvalidType userInfo:@{NSLocalizedDescriptionKey: exception.reason}];
```

Comment:
@mattheworiordan I decided to not use the `isValidJSONObject` method because it doesn't tell us why it isn't valid. On the other hand, catching the exception gives us more details about the failure.